# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def addTwo(x):
    return x + 2

print(addTwo(2))

print((lambda x: x + 2)(2))

4
4


#### Saving to a Variable

In [5]:
add_two = lambda x: x + 2
print(add_two(2))

4


#### Multiple Inputs

In [7]:
print((lambda x, y, z: x * y * z)(3, 5, 8))

multiply_three = lambda x, y, z: x * y * z
print(multiply_three(3, 5, 8))

120
120


#### Passing a Lambda into a Function

In [8]:
def multiply(f, num):
    # f will be a lambda
    # num will be a number
    return f(num)

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [14]:
def multiply_test(num):
    return num * 4

def returnFunc():
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(f_return(5))

f_return = returnFunc()(5)
print(f_return)

def returnLam(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4, 5)
print(r_lamb(6,4))

print(returnLam(4, 5)(6, 4))
print(returnLam(2, 5)(10, 4))

10
10
19
19
21


#### If Statements within Lambdas

In [15]:
print((lambda num: num * 2 if num > 10 else num + 2)(9))

f_condition = lambda num: num * 2 if num > 2 else num + 2
print(f_condition(11))

11
22


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [25]:
def cube():
    return lambda x: x ** 3
f = cube()
print(f(4))

64
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [32]:
def squared(num1, num2):
    if num1 < 10 > num2:
        return num1**2, num2**2
    return num1, num2

numbers = [4, 11, 20, 3, 15, 20]
numbers2 = [4, 10, 3, 2, 6]

print(list(map(squared, numbers, numbers2)))
print([t for t in map(squared, numbers, numbers2)])

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [34]:
double_num = print(list(map(lambda num1, num2: (num1**2, num2**2) if num1 < 10 > num2 else (num1, num2), numbers, numbers2)))

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


In [35]:
square_num = print(list(map(lambda x: x**2, [10, 1, 3, 9])))

[100, 1, 9, 81]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [43]:
alist = [10, 1, 3, 9, 8]
quick_maths = print(list(map(lambda x: x * 2 - 1, alist)))

[19, 1, 5, 17, 15]


##### Use the map function to take in a list of numbers and make each index False if odd, else True if even

In [42]:
alist = [10, 1, 3, 9, 8]
find_evens = print(list(map(lambda x: True if x%2 == 0 else False, alist)))

[True, False, False, False, True]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [3]:
names = ['mike', 'brendan', 'james', 'jordan', 'hayden']

def filterj(alist):
    nojs = []
    for name in alist:
        if name[0] == 'j':
            continue
        nojs.append(name.title())
    return nojs

print(filterj(names))

['Mike', 'Brendan', 'Hayden']


#### Using Lambda's with Filter()

In [11]:
print(list(filter(lambda name: True if name[0] != 'j' else False, names)))

print([n for n in filter(lambda name: True if name[0] != 'j' else False, names)])

['mike', 'brendan', 'hayden']
['mike', 'brendan', 'hayden']


In [14]:
def filter_copy(f, alist):
    copy_list = []
    for i in range(len(alist)):
        if f(alist[i]):
            copy_list.append(alist[i])
    return copy_list

print(filter_copy(lambda x: True if x % 2 == 0 else False, [1, 2, 3, 4, 5]))

print(list(filter(lambda x: True if x % 2 == 0 else False, [1, 2, 3, 4, 5])))

[2, 4]
[2, 4]


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [20]:
import statistics
# x = statistics.mean(data1)

nums = [2,7,4.2,1.6,9,4.4,4.9]
# 4.729

nums2 = [-4.5, -2, 2.2, 3, 4, 5, 7.8]
# 2.214

# mean = statistics.mean(nums)
# print(list(filter(lambda x: True if x < mean else False, nums)))

print(list(filter(lambda x: True if x < statistics.mean(nums) else False, nums)))
print(list(filter(lambda x: True if x < statistics.mean(nums2) else False, nums2)))


[2, 4.2, 1.6, 4.4]
[-4.5, -2, 2.2]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [22]:
from functools import reduce

list1 = [2, 4, 6, 8, 10]

def addNums(nums1, nums2):
    return nums1 + nums2

print(reduce(addNums, list1))

30


#### Using Lambda's with Reduce()

In [23]:
print(reduce(lambda x, y: x-y, list1))

-26


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [28]:
my_list = [1, 2, 3, 4]

print(reduce(lambda x, y: x * y, my_list))

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [31]:
def recursive_addition(num):
    if num <= 1:
        print(f'r_a({num}) = 1')
        return num
    print(f'r_a({num}) = {num} + r_a({num-1})')
    return num + recursive_addition(num-1)

print(recursive_addition(5))

# r**2-x
# 2*(r**2-(r-1)**2)-r
# 2*(49-36)-7

# 8   36=8+28
# 7   28=7+21
# 6   21=6+15
# 5   15=5+10
# 4   10=4+6
# 3   6=3+3
# 2   3=2+1
# 1   1

r_a(5) = 5 + r_a(4)
r_a(4) = 4 + r_a(3)
r_a(3) = 3 + r_a(2)
r_a(2) = 2 + r_a(1)
r_a(1) = 1
15


#### Writing a Factorial Function

In [34]:
def facto(num):
    if num <= 1:
        print(f'facto({num}) = 1')
        return num
    print(f'facto({num}) = {num} * facto({num-1})')
    return num * facto(num-1)

print(facto(5))

facto(5) = 5 * facto(4)
facto(4) = 4 * facto(3)
facto(3) = 3 * facto(2)
facto(2) = 2 * facto(1)
facto(1) = 1
120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [40]:
def rec_sub(num):
    if num <= 1:
        return num
    return num - rec_sub(num-1)

print(rec_sub(15))

# 8   4 = 8-4
# 7   4 = 7-3
# 6   3 = 6-3
# 5   3 = 5-2
# 4   2 = 4-2
# 3   2 = 3-1
# 2   1 = 2-1
# 1   1

8


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [42]:
def my_range(start, stop, step=1):
    while start < stop:
        yield start
        start += step

print([num for num in my_range(2, 20, 2)])

[2, 4, 6, 8, 10, 12, 14, 16, 18]
<generator object my_range at 0x000001BF5559BE60>


In [44]:
def my_range(start, stop, step=1):
    while start < stop:
        yield start
        start += step

print([num for num in my_range(2, 20, 2)])

2


#### Infinite Generator

In [ ]:
# bad, never create infinite loops
def my_range(start, stop):
    while start < stop:
        yield start

print([num for num in my_range(2, 20, 2)])

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [5]:
def square_down(start, stop, step=1):
    while start > stop:
        yield start ** 2
        start -= step

print([num for num in square_down(10, 1, 2)])

[100, 64, 36, 16, 4]


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [11]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

import re
no_space = re.compile(r'[\w]+')

# for n in places:
#     if no_space.search(n):
#         print(n)

print(list(filter(lambda x: True if no_space.search(x) else False, places)))

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [21]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

import re
last = re.compile(r'([A-Za-z]+) ([A-Za-z]+)* ([a-zA-Z]+)')

# for n in author:
#     found = last.search(n)
#     if found:
#         print(found.groups())

print(author.sort(key=lambda x: len(x)))

SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='? (1323453998.py, line 11)

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [42]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

print(list(map(lambda x: (x[0], x[-1] * 9/5 + 32), places)))

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [10]:
# def recursive_addition(num):
#     if num <= 1:
#         print(f'r_a({num}) = 1')
#         return num
#     print(f'r_a({num}) = {num} + r_a({num-1})')
#     return num + recursive_addition(num-1)

# print(recursive_addition(5))

# I know assignment says fib(5) should give 8, but math sites online seemed to set fib(5) = 5 which is 1 iteration behind the given output for this exercise.

def fib(n):
    if n <= 1:
        # print(f'fib({n}) = {n}')
        return n
    # print(f'fib({n}) = fib({n-1}) + fib({n-2})')
    return fib(n-1) + fib(n-2)

print(fib(0))
print(fib(1))
print(fib(2))
print(fib(3))
print(fib(4))
print(fib(5))
print(fib(6))

0
1
1
2
3
5
8
